### PR TITLE
Offline icons

### PR DESCRIPTION
**Pull request by adambkaplan**
Preview support for offline icons in the App CRD.

Idea is to enable this feature based on the direction of [operator-lifecycle-manager](https://github.com/operator-framework/operator-lifecycle-manager):

1. Define icon inline as binary data (jpeg, png) or text (svg)
2. Alternatively, allow references to a ConfigMap with contents (text or binary with kube 1.10)

### CODE REVIEW

File: pkg/apis/app/v1beta1/zz_generated.kubebuilder.go [hunk 1]
```diff
@@ -230,40 +268,124 @@ var (
...
+																},
+																"fieldPath": v1beta1.JSONSchemaProps{
+																	Type: "string",
+																},
 																"key": v1beta1.JSONSchemaProps{
 																	Type: "string",
 																},
+																"kind": v1beta1.JSONSchemaProps{
+																	Type: "string",
+																},
```
Comment (adambkaplan) [inline pkg/apis/app/v1beta1/zz_generated.kubebuilder.go:277]: Note - generated from kubebuilder. Don't think folks have been diligent re: generation

File: pkg/apis/app/v1beta1/application_types.go [hunk 2]
```diff
@@ -84,15 +84,34 @@ type ApplicationStatus struct {
...
+	//
+	// Only one of Source or Data may be defined.
+	Source string `json:"src,omitempty"`
 
-	// (optional) The size of the image in pixels (e.g., 25x25).
+	// Size of the image in pixels (e.g., 25x25). Optional.
 	Size string `json:"size,omitempty"`
 
-	// (optional) The mine type of the image (e.g., "image/png").
+	// Type indicates the MIME type of the image (e.g., "image/png"). Optinal.
```
Comment (bparees) [inline pkg/apis/app/v1beta1/application_types.go]: typo

File: pkg/apis/app/v1beta1/application_types.go [hunk 3]
```diff
@@ -84,15 +84,34 @@ type ApplicationStatus struct {
...
+// ImageData contains raw information for the image, either as binary, text, or a reference to a ConfigMap entry.
+type ImageData struct {
+	// Binary contains the base64-encoded image binary.
+	Binary string `json:"binary,omitempty"`
+
+	// Text contains the image definition as plain text, for formats such as SVG.
+	Text string `json:"text,omitempty"`
+
+	// ConfigMapKeyRef is a reference to a ConfigMap entry containing the image data.
+	ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty"`
```
Comment (bparees) [inline pkg/apis/app/v1beta1/application_types.go]: why not just move the "source url" in here too?  You still only want exactly one of these to be filled in.

```
type Image struct {
  Binary string
  Text string
  URL string
  ConfigMapKeyRef *ConfigMapKeySelector
}
```

Comment (adambkaplan) [reply pkg/apis/app/v1beta1/application_types.go]: Certainly doable. I was going for a syntax along the lines of EnvVar in the ContainerSpec. Curious if this syntax remains in the v2 API work.

I'd probably keep `Source`, though, as this has merged and therefore "landed" as the API.

Comment (bparees) [reply pkg/apis/app/v1beta1/application_types.go]: it's a beta, i don't think it having landed is a good reason to feel like we're stuck with it if there is agreement on a better approach before it becomes finalized.

**Comment by adambkaplan** [comment]
Moved to https://github.com/kubernetes-sigs/application/pull/58

**Pull request closed without merging**